File: androidApp/src/main/kotlin/com/malaria/android/data/entities/AuditEntry.kt
package com.malaria.android.data.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import kotlinx.datetime.Instant
import java.util.UUID

/**
 * Append-only audit event (spec §8). The DAO surface intentionally exposes
 * no update or delete methods; append-only is an application convention
 * mirrored from `iosApp/Models/AuditEntry.swift`.
 *
 * `seq` is a monotonic per-device sequence assigned by [AuditDao.write] as
 * `max(seq) + 1` inside a transaction. SQLite AUTOINCREMENT was rejected
 * because the iOS SwiftData side computes seq the same way; cross-platform
 * comparability requires identical semantics.
 *
 * Column names are snake_case so the schema-drift JSON export has a stable
 * surface to compare against `docs/SCHEMA.md`.
 */
@Entity(
    tableName = "audit_entries",
    indices = [
        Index("seq", unique = true),
        Index("action"),
        Index("timestamp"),
    ],
)
data class AuditEntry(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String = UUID.randomUUID().toString(),

    @ColumnInfo(name = "seq")
    val seq: Long,

    @ColumnInfo(name = "timestamp")
    val timestamp: Instant,

    @ColumnInfo(name = "actor_id")
    val actorId: String,

    @ColumnInfo(name = "actor_role_at_time")
    val actorRoleAtTime: String,

    @ColumnInfo(name = "action")
    val action: String,

    @ColumnInfo(name = "resource_type")
    val resourceType: String? = null,

    @ColumnInfo(name = "resource_id")
    val resourceId: String? = null,

    @ColumnInfo(name = "metadata_json")
    val metadataJson: String = "{}",

    @ColumnInfo(name = "override_context")
    val overrideContext: String? = null,

    @ColumnInfo(name = "override_reason")
    val overrideReason: String? = null,

    @ColumnInfo(name = "override_notes")
    val overrideNotes: String? = null,

    @ColumnInfo(name = "context_reviewed")
    val contextReviewed: Boolean? = null,

    @ColumnInfo(name = "override_actor_initials")
    val overrideActorInitials: String? = null,

    @ColumnInfo(name = "app_version")
    val appVersion: String,

    @ColumnInfo(name = "os_version")
    val osVersion: String,
)

File: iosApp/Models/AuditEntry.swift
import Foundation
import SwiftData

/// Append-only audit event. The repository layer must never expose update
/// or delete operations on this type (spec §8). `seq` is a monotonic
/// per-device sequence assigned by `AuditRepository.write` at insert time.
@Model
final class AuditEntry {
    @Attribute(.unique) var id: String
    var seq: Int
    var timestamp: Date
    var actorId: String
    var actorRoleAtTime: String
    var action: String
    var resourceType: String?
    var resourceId: String?
    var metadataJson: String
    var overrideContext: String?
    var overrideReason: String?
    var overrideNotes: String?
    var contextReviewed: Bool?
    var overrideActorInitials: String?
    var appVersion: String
    var osVersion: String

    init(
        id: String = UUID().uuidString,
        seq: Int,
        timestamp: Date,
        actorId: String,
        actorRoleAtTime: String,
        action: String,
        resourceType: String? = nil,
        resourceId: String? = nil,
        metadataJson: String = "{}",
        overrideContext: String? = nil,
        overrideReason: String? = nil,
        overrideNotes: String? = nil,
        contextReviewed: Bool? = nil,
        overrideActorInitials: String? = nil,
        appVersion: String,
        osVersion: String
    ) {
        self.id = id
        self.seq = seq
        self.timestamp = timestamp
        self.actorId = actorId
        self.actorRoleAtTime = actorRoleAtTime
        self.action = action
        self.resourceType = resourceType
        self.resourceId = resourceId
        self.metadataJson = metadataJson
        self.overrideContext = overrideContext
        self.overrideReason = overrideReason
        self.overrideNotes = overrideNotes
        self.contextReviewed = contextReviewed
        self.overrideActorInitials = overrideActorInitials
        self.appVersion = appVersion
        self.osVersion = osVersion
    }
}
